## Clustering draft 1

**General idea** : test the functions defined in the module1.py file and make a first application.

**General plan** : 
1. Import the data using the module YahooFinance 

2. Repeat the clustering several times to get several clustering outcomes

3. Compute the weights of each stocks within each cluster 

4. Compute the returns of each portfolio, where one portfolio corresponds to one clustering and is composed of 5 big assets (which are the 5 clusters) **in this notebook, we first fix the number of clusters to be equal to 5**

Here are the main package we use in this notebook

In [4]:
import pandas as pd
import numpy as np
import yfinance as yf
from sklearn.pipeline import Pipeline
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from bs4 import BeautifulSoup
import requests 
from pypfopt.efficient_frontier import EfficientFrontier


We also import the module1.py file to use the function we defined there.

In [5]:
pip install git+https://github.com/robertmartin8/PyPortfolioOpt.git

  Cloning https://github.com/robertmartin8/PyPortfolioOpt.git to /tmp/pip-req-build-h_9vtjpy
  Running command git clone --filter=blob:none --quiet https://github.com/robertmartin8/PyPortfolioOpt.git /tmp/pip-req-build-h_9vtjpy
  Resolved https://github.com/robertmartin8/PyPortfolioOpt.git to commit 30ab57147ba61eddc8301294a5b1c5ef260b23fa
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Note: you may need to restart the kernel to use updated packages.


In [6]:
## we make some manipulations to correctly import module1

import os
import sys

# on a obtenu le chemin absolu vers notre répertoire en utilisant
# le terminal (ls, pwd, cd)
module_1_directory = '/Users/khelifanail/Documents/GitHub/Portfolio_clustering_project/code'

# Ajouter le chemin du répertoire 'code' au chemin de recherche Python
sys.path.append(module_1_directory)

# Maintenant, vous pouvez importer module1
import module1

### 1. Importation of data

We download various types of data (Open, High, Low, Close, Adj Close, Volume) for **496** assets between two periods (start, end) into a pandas dataframe denoted as df.

We then store the returns of each of these assets in the 'data'.

#### 1.1. Scraping the names of the assets of interest

- This is important because to import financial data for assets on Yahoo Finance, you need the tickers' names. This scraping work was performed in the file *get_S&P500_symbols.ipynb* and generated the *S&P500_symbols.csv* file.

- We rely on the csv package to read the *S&P500_symbols.csv* file and to create a list containing all the tickers

In [7]:
import csv 

with open('S&P500_symbols.csv', mode='r') as file:
    # Create a CSV reader
    csv_reader = csv.reader(file)
    
    # Transform the CSV data into a list
    stock_symbols = []
    for row in csv_reader:
        stock_symbols.append(row[0])

# Close the file
file.close()

stock_symbols.pop(0)

'Symbols of S&P 500'

In [8]:
stock_symbols

['MMM',
 'AOS',
 'ABT',
 'ABBV',
 'ACN',
 'ADM',
 'ADBE',
 'ADP',
 'AES',
 'AFL',
 'A',
 'ABNB',
 'APD',
 'AKAM',
 'ALK',
 'ALB',
 'ARE',
 'ALGN',
 'ALLE',
 'LNT',
 'ALL',
 'GOOGL',
 'GOOG',
 'MO',
 'AMZN',
 'AMCR',
 'AMD',
 'AEE',
 'AAL',
 'AEP',
 'AXP',
 'AIG',
 'AMT',
 'AWK',
 'AMP',
 'AME',
 'AMGN',
 'APH',
 'ADI',
 'ANSS',
 'AON',
 'APA',
 'AAPL',
 'AMAT',
 'APTV',
 'ACGL',
 'ANET',
 'AJG',
 'AIZ',
 'T',
 'ATO',
 'ADSK',
 'AZO',
 'AVB',
 'AVY',
 'AXON',
 'BKR',
 'BALL',
 'BAC',
 'BBWI',
 'BAX',
 'BDX',
 'WRB',
 'BRK.B',
 'BBY',
 'BIO',
 'TECH',
 'BIIB',
 'BLK',
 'BX',
 'BK',
 'BA',
 'BKNG',
 'BWA',
 'BXP',
 'BSX',
 'BMY',
 'AVGO',
 'BR',
 'BRO',
 'BF.B',
 'BG',
 'CHRW',
 'CDNS',
 'CZR',
 'CPT',
 'CPB',
 'COF',
 'CAH',
 'KMX',
 'CCL',
 'CARR',
 'CTLT',
 'CAT',
 'CBOE',
 'CBRE',
 'CDW',
 'CE',
 'COR',
 'CNC',
 'CNP',
 'CDAY',
 'CF',
 'CRL',
 'SCHW',
 'CHTR',
 'CVX',
 'CMG',
 'CB',
 'CHD',
 'CI',
 'CINF',
 'CTAS',
 'CSCO',
 'C',
 'CFG',
 'CLX',
 'CME',
 'CMS',
 'KO',
 'CTSH',
 'CL',


#### 1.2. Creating the dataframe using pandas and yfinance 

In this case, we focus on the financial data of asset during the year 2022

In [9]:
n_stocks = len(stock_symbols) # number of stocks = 502

start = "2022-01-01" # start date
end = "2022-12-01" # end date

df = pd.DataFrame(yf.download(stock_symbols, start, end)) # data on the 198 assets
data = np.log(df["Close"]/df["Open"]).transpose() # compute the returns of these assets
data = data.dropna()



[*********************100%%**********************]  502 of 502 completed


10 Failed downloads:
['MHK', 'CSGP', 'WST', 'FI', 'JBHT']: OperationalError('database is locked')
['VLTO', 'KVUE', 'GEHC']: Exception("%ticker%: Data doesn't exist for startDate = 1641013200, endDate = 1669870800")
['BRK.B']: Exception('%ticker%: No timezone found, symbol may be delisted')
['BF.B']: Exception('%ticker%: No price data found, symbol may be delisted (1d 2022-01-01 -> 2022-12-01)')


In [10]:
data

Date  2022-01-03  2022-01-04  2022-01-05  2022-01-06  2022-01-07  2022-01-10  \
A      -0.015976   -0.028044   -0.014895    0.001812   -0.026984    0.012966   
AAL     0.028125   -0.010460   -0.023804   -0.017085    0.027875   -0.023667   
AAPL    0.023234   -0.016173   -0.026459   -0.004061   -0.004173    0.018227   
ABBV    0.000074   -0.001257    0.006424   -0.008615   -0.002739    0.009651   
ABNB    0.043696   -0.019447   -0.045972   -0.003375    0.042253    0.005929   
...          ...         ...         ...         ...         ...         ...   
YUM    -0.013459    0.000583   -0.014505    0.006742   -0.007434   -0.009694   
ZBH     0.013880    0.011876   -0.008349   -0.015088    0.007441   -0.003461   
ZBRA   -0.013912    0.004862   -0.042476    0.000288   -0.039303    0.015130   
ZION    0.005620   -0.000605   -0.003768    0.027053    0.013068   -0.015973   
ZTS    -0.034686   -0.038773   -0.047082    0.008963   -0.028212    0.021298   

Date  2022-01-11  2022-01-12  2022-01-13  2022-01-14  ...  2022-11-16  \
A       0.011247    0.011503   -0.029123    0.004433  ...   -0.007247   
AAL     0.007388   -0.035054    0.032049   -0.032982  ...   -0.025229   
AAPL    0.015890   -0.003356   -0.020635    0.010046  ...   -0.002283   
ABBV    0.008211    0.005624   -0.011023    0.018721  ...   -0.010285   
ABNB    0.047987   -0.004120   -0.025180   -0.010675  ...   -0.030833   
...          ...         ...         ...         ...  ...         ...   
YUM    -0.018652   -0.015062   -0.003719   -0.008308  ...    0.004771   
ZBH     0.006329   -0.019556    0.014868    0.004169  ...   -0.006097   
ZBRA    0.006370   -0.002837   -0.026315    0.014806  ...   -0.005133   
ZION    0.001425   -0.004547   -0.005551    0.014112  ...   -0.022113   
ZTS     0.003766   -0.015064   -0.026082   -0.003679  ...   -0.010265   

Date  2022-11-17  2022-11-18  2022-11-21  2022-11-22  2022-11-23  2022-11-25  \
A      -0.006143   -0.003346   -0.003302    0.031605   -0.008461    0.011985   
AAL     0.009363   -0.014836   -0.012200    0.009342    0.030988    0.006920   
AAPL    0.028876   -0.006719   -0.014422    0.013744    0.010781   -0.001349   
ABBV    0.007851    0.008489    0.010944    0.013229   -0.003507    0.000188   
ABNB   -0.025159   -0.030917   -0.013491    0.018538    0.011135    0.007193   
...          ...         ...         ...         ...         ...         ...   
YUM    -0.000651    0.003054   -0.005789    0.003930    0.002624    0.005309   
ZBH    -0.005603    0.004216    0.009739    0.007048    0.013995    0.006396   
ZBRA    0.013099   -0.012765    0.011939    0.014081    0.029699    0.005162   
ZION   -0.004521   -0.021279    0.000985    0.001562    0.006817    0.004066   
ZTS    -0.020676   -0.003359    0.013668    0.003640    0.011833   -0.002462   

Date  2022-11-28  2022-11-29  2022-11-30  
A      -0.023170   -0.006208    0.023503  
AAL    -0.036213    0.017839    0.021717  
AAPL   -0.006359   -0.021860    0.045822  
ABBV   -0.009861   -0.000695    0.014939  
ABNB   -0.017681   -0.007520    0.063456  
...          ...         ...         ...  
YUM     0.001345   -0.000317    0.018276  
ZBH    -0.003986    0.015583    0.005176  
ZBRA    0.002307   -0.011070    0.013634  
ZION   -0.022052    0.012026    0.020077  
ZTS    -0.008939    0.001896    0.040041  

[491 rows x 230 columns]

Due to missing values in the data, 5 stocks have been excluded from the intial 502 we considered. Thus the shape of the dataframe is (496, 230) 






### 2. We repeat the clustering step several times to start preparing multiple micro-portfolios

**General idea**: The multiple_clustering function takes as arguments:

1. *n_repeat*: an integer that corresponds to the number of times we want to train (.fit()) the model passed as an argument on the data.

2. *data*: a pandas dataframe (like the one we generated above) that corresponds to the financial data on which the model will be trained.

3. *model*: a sklearn clustering model (for now, we will test it with KMeans).

4. *model_name*: the name of the model, necessary for creating the pipeline properly.

**Outcome**: a pandas dataframe storing the results of the multiple clustering and a dictionary containing the centroids of the clusters each time 

We first make the test using the K-Means model and fix we fix the number of clusters to 5

In [7]:
model = KMeans(n_clusters=5)
model_name = 'kmeans'

Y, C = module1.multiple_clusterings(10, data, model, model_name)
print(Y, C)

/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` ex

      Clustering n°1  Clustering n°2  Clustering n°3  Clustering n°4  \
A                  1               1               3               2   
AAL                0               4               0               0   
AAPL               1               1               3               2   
ABBV               3               3               1               4   
ABNB               0               4               0               0   
...              ...             ...             ...             ...   
YUM                1               1               3               2   
ZBH                1               1               3               2   
ZBRA               4               1               4               1   
ZION               1               1               3               2   
ZTS                1               1               3               2   

      Clustering n°5  Clustering n°6  Clustering n°7  Clustering n°8  \
A                  0               1               0           

/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
/usr/local/lib/python3.11/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` ex

In [8]:
Y

Clustering n°1  Clustering n°2  Clustering n°3  Clustering n°4  \
A                  1               1               3               2   
AAL                0               4               0               0   
AAPL               1               1               3               2   
ABBV               3               3               1               4   
ABNB               0               4               0               0   
...              ...             ...             ...             ...   
YUM                1               1               3               2   
ZBH                1               1               3               2   
ZBRA               4               1               4               1   
ZION               1               1               3               2   
ZTS                1               1               3               2   

      Clustering n°5  Clustering n°6  Clustering n°7  Clustering n°8  \
A                  0               1               0               0   
AAL                2               3               1               2   
AAPL               3               1               0               0   
ABBV               1               2               4               1   
ABNB               2               3               1               4   
...              ...             ...             ...             ...   
YUM                1               1               0               0   
ZBH                0               4               0               0   
ZBRA               3               1               2               0   
ZION               0               4               0               0   
ZTS                1               1               0               0   

      Clustering n°9  Clustering n°10  
A                  2                3  
AAL                3                0  
AAPL               2                3  
ABBV               0                4  
ABNB               3                0  
...              ...              ...  
YUM                0                3  
ZBH                2                3  
ZBRA               4                0  
ZION               2                1  
ZTS                2                3  

[496 rows x 10 columns]

### 3. Compute the weights of the stocks in each cluster 

#### 3.1. Get the cluster compositions each time 

**Why doing so?**: the purpose of this work is to compute the returns of each cluster (seen as a new "fictive" asset composed of "real" assets) and this latter task is easier if we know the cluster composition in terms of tickers (because in the financial database we have data given a ticker).

In [9]:
Y_symbol = module1.cluster_composition(Y)
Y_symbol

Clustering n°1  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°2  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, BX, CCL...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°3  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°4  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°5  \
Cluster 1  [A, ACN, AIG, ALLE, AMP, AOS, APH, APTV, AVY, ...   
Cluster 2  [AAL, ABNB, ALK, BA, BBWI, CCL, CZR, DAL, EXPE...   
Cluster 3  [AAPL, ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, ...   
Cluster 4  [ABBV, ABT, ACGL, ADM, ADP, AEE, AEP, AES, AFL...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°6  \
Cluster 1  [A, AAPL, ABT, ACN, ADI, ADP, AES, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ADBE, ADSK, ALB, ALGN, AMAT, AMD, ...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [AIG, ALK, AMP, AVY, AXP, BA, BAC, BALL, BBY, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°7  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, AJG, ALL...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°8  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AJG, AKAM, ...   
Cluster 2  [AAL, ALK, APTV, BA, BBWI, BBY, BKNG, BWA, BX,...   
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, ALL, AMC...   
Cluster 4  [ABNB, ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                              Clustering n°9  \
Cluster 1  [A, AAPL, ABT, ACN, AES, AIG, AKAM, ALLE, AME,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   
Cluster 3  [ABBV, ACGL, ADM, ADP, AEE, AEP, AFL, AIZ, AJG...   
Cluster 4  [ADBE, ADI, ADSK, ALB, ALGN, AMAT, AMD, AMZN, ...   
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...   

                                             Clustering n°10  
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AJG, AKAM, ALLE,...  
Cluster 2  [AAL, ABNB, ADBE, ADI, ADSK, ALB, ALGN, AMAT, ...  
Cluster 3  [ABBV, ACGL, ADM, AEE, AEP, AFL, AIZ, ALL, AMC...  
Cluster 4  [AIG, ALK, AMP, AVY, AXP, BA, BAC, BBY, BEN, B...  
Cluster 5  [APA, BKR, CF, COP, CTRA, CVX, DVN, EOG, EQT, ...

In [10]:
Y_symbol.columns

Index(['Clustering n°1', 'Clustering n°2', 'Clustering n°3', 'Clustering n°4',
       'Clustering n°5', 'Clustering n°6', 'Clustering n°7', 'Clustering n°8',
       'Clustering n°9', 'Clustering n°10'],
      dtype='object')

#### 3.2. Compute the weights of each assets within a cluster

**Why doing so?**: because each stock in a cluster has a different weight. Stock are "representative" of the cluster in which they live to a different extent. So far, we built the weight of a stock based on its $\mathcal{L}^2$ distance from the cluster's centroid. We can therefore see the weights as a kind of "inertia contribution" of the stock in the cluster.

##### 3.2.1. Analysis of the centroids 

- The .multiple_clusterings() function returns two data frame, the first of which is a dataframe containing the centroid of each cluster. This centroid is an array containing the "average" return each day, where each daily average is taken on the returns of all the stocks contained in the corresponding cluster. 

- Here is an example:


In [11]:
C

Clustering n°1  \
Cluster 1  [1.1244838244132374, -0.2749346042866737, -0.8...   
Cluster 2  [-0.19806314356161475, 0.20159348632009083, -0...   
Cluster 3  [1.5371658112463586, 1.1520938247611001, -0.38...   
Cluster 4  [-0.043873467042549096, -0.014869674776017127,...   
Cluster 5  [-0.17409859643615738, -1.138527299010478, -1....   

                                              Clustering n°2  \
Cluster 1  [-0.0625870737441699, -1.1188149295966316, -1....   
Cluster 2  [-0.2286692164407662, 0.15512644818402174, -0....   
Cluster 3  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 4  [-0.04351084201028617, -0.013742211374664193, ...   
Cluster 5  [0.9978838118027735, -0.14563325203064836, -0....   

                                              Clustering n°3  \
Cluster 1  [1.1301311810523154, -0.11219298097335409, -0....   
Cluster 2  [-0.05479170637683879, -0.022171875492677098, ...   
Cluster 3  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 4  [-0.2060873518326184, 0.1896518623420935, 0.00...   
Cluster 5  [-0.17409859643615738, -1.138527299010478, -1....   

                                              Clustering n°4  \
Cluster 1  [1.1301311810523154, -0.11219298097335409, -0....   
Cluster 2  [-0.25594876226768776, -1.1625746749393877, -1...   
Cluster 3  [-0.18158350297936524, 0.22692225208571215, 0....   
Cluster 4  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 5  [-0.058397438328084104, -0.01778263123422645, ...   

                                              Clustering n°5  \
Cluster 1  [-0.056468189874353085, 0.3559927835110836, 0....   
Cluster 2  [-0.18022609311933518, -0.06324551173272618, 0...   
Cluster 3  [1.3182102697427653, -0.5985279028775956, -0.9...   
Cluster 4  [-0.21357728725034417, -1.1564066500308154, -1...   
Cluster 5  [1.5002065476511288, 1.0570143644422916, -0.31...   

                                              Clustering n°6  \
Cluster 1  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 2  [-0.6950392404750687, -0.4195327214725122, -0....   
Cluster 3  [-0.025196306740732404, 0.02789418507861987, 0...   
Cluster 4  [0.168969578999866, -1.09751363474114, -1.2008...   
Cluster 5  [0.3696676080372728, 0.7112020020605385, 0.210...   

                                              Clustering n°7  \
Cluster 1  [-0.16674186263899254, 0.2411596799902072, 0.0...   
Cluster 2  [1.1273320614067666, -0.2772005302773248, -0.8...   
Cluster 3  [-0.23752576235501585, -1.158431753465747, -1....   
Cluster 4  [1.5293699302639876, 1.106898060194132, -0.357...   
Cluster 5  [-0.06246901607733254, -0.038960105583582476, ...   

                                              Clustering n°8  \
Cluster 1  [-0.269961026745664, 0.11509131685467602, 0.02...   
Cluster 2  [-0.011407222323663509, 0.01921203823905055, 0...   
Cluster 3  [0.6948348395841755, 0.07769453282369482, -0.6...   
Cluster 4  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 5  [-0.017598589793172286, -1.1204793178796761, -...   

                                              Clustering n°9  \
Cluster 1  [-0.09675130333603825, -0.05251703469421325, 0...   
Cluster 2  [1.5002065476511288, 1.0570143644422916, -0.31...   
Cluster 3  [-0.1741411600739638, 0.2528083181597799, 0.03...   
Cluster 4  [1.0818821358026836, -0.10293663224269473, -0....   
Cluster 5  [-0.2096597280554139, -1.1490651258162525, -1....   

                                             Clustering n°10  
Cluster 1  [0.12862317441124896, -1.0424515047200198, -1....  
Cluster 2  [0.25570898346138965, 0.6440613100704217, 0.17...  
Cluster 3  [1.5293699302639876, 1.106898060194132, -0.357...  
Cluster 4  [-0.6376964196309345, -0.38199669770814304, -0...  
Cluster 5  [0.02584249288452328, 0.053727821859144716, 0....

##### 3.2.2. Computing the weights of each cluster

- In module1, there are two functions *cluster_weights* and *gaussian_weights* which, given one cluster and its centroids, return the weights of the stocks in the cluster. These weights corresponds to a sort of distance of the stocks to the cluster's centroid.

- If we consider the previous example, we can consider the first cluster (Cluster 1 of the Clusering n°1) and compute the weights of the stocks in this cluster. 


In [12]:
Y_symbol.head(2)

Clustering n°1  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   

                                              Clustering n°2  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, BX, CCL...   

                                              Clustering n°3  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   

                                              Clustering n°4  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   

                                              Clustering n°5  \
Cluster 1  [A, ACN, AIG, ALLE, AMP, AOS, APH, APTV, AVY, ...   
Cluster 2  [AAL, ABNB, ALK, BA, BBWI, CCL, CZR, DAL, EXPE...   

                                              Clustering n°6  \
Cluster 1  [A, AAPL, ABT, ACN, ADI, ADP, AES, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ADBE, ADSK, ALB, ALGN, AMAT, AMD, ...   

                                              Clustering n°7  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AKAM, ALLE,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   

                                              Clustering n°8  \
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AIG, AJG, AKAM, ...   
Cluster 2  [AAL, ALK, APTV, BA, BBWI, BBY, BKNG, BWA, BX,...   

                                              Clustering n°9  \
Cluster 1  [A, AAPL, ABT, ACN, AES, AIG, AKAM, ALLE, AME,...   
Cluster 2  [AAL, ABNB, ALK, APTV, BA, BBWI, BKNG, CCL, CZ...   

                                             Clustering n°10  
Cluster 1  [A, AAPL, ABT, ACN, ADP, AES, AJG, AKAM, ALLE,...  
Cluster 2  [AAL, ABNB, ADBE, ADI, ADSK, ALB, ALGN, AMAT, ...

#### **Example**: 
- To take a basic example, we consider the first cluster of the first clustering (if we refer to the Y_symbol cell above this corresponds to *Y_symbol.iloc[0, 0]* for the tickers and to *C.iloc[0, 0]* for the centroids).

- We then compute the $\mathcal{L}^2$ weights and the Gaussian weights for this cluster 



In [13]:
    ## We consider the first cluster of the first clustering 
    #  (location [0, 0])

cluster = Y_symbol.iloc[0,0]
centroid = C.iloc[0,0]

    ## We then compute the weights (L2 and Gaussian) corresponding to 
    #  the stocks in this cluster 

weights_L2 = module1.cluster_weights(cluster, centroid, data)
weights_gaussian = module1.gaussian_weights(cluster, centroid, data)


In [14]:
cluster_composition = cluster 
micro_portfolio_return = pd.DataFrame(index=cluster_composition, columns=data.columns).transpose()
micro_portfolio_return

A AAPL  ABT  ACN  ADP  AES  AIG AKAM ALLE  AME  ...  WRK  WST  \
Date                                                          ...             
2022-01-03  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-01-04  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-01-05  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-01-06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-01-07  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
...         ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
2022-11-23  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-11-25  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-11-28  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-11-29  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
2022-11-30  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   

            WTW   WY XRAY  XYL  YUM  ZBH ZION  ZTS  
Date                                                
2022-01-03  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-01-04  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-01-05  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-01-06  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-01-07  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...         ...  ...  ...  ...  ...  ...  ...  ...  
2022-11-23  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-11-25  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-11-28  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-11-29  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2022-11-30  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[230 rows x 236 columns]

- We make sure that the weights add up to one

In [15]:
print(weights_L2.sum(axis=1))
print(weights_gaussian.sum(axis=1))

0    1.0
dtype: float64
0    1.0
dtype: float64


In [32]:
weights_gaussian

0        1         2         3         4         5        6    \
0  0.000125  0.00101  0.000076  0.000307  0.000073  0.000957  0.00195   

       7         8         9    ...       226       227       228       229  \
0  0.00018  0.000184  0.000074  ...  0.001662  0.000059  0.000171  0.000609   

        230       231       232      233       234      235  
0  0.000251  0.000143  0.000232  0.00086  0.003765  0.00003  

[1 rows x 236 columns]

In [37]:
weights_gaussian.applymap(lambda x: x if x > 0.003 else None)

/var/folders/q2/wg5gyfhj2r9cd97zfmckktvw0000gn/T/ipykernel_47509/803272374.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weights_gaussian.applymap(lambda x: x if x > 0.003 else None)


0     1     2     3     4     5     6     7     8     9    ...   226  \
0  None  None  None  None  None  None  None  None  None  None  ...  None   

    227   228   229   230   231   232   233       234   235  
0  None  None  None  None  None  None  None  0.003765  None  

[1 rows x 236 columns]

### 4. Compute the returns of each portfolio, 

- **Idead**: We considered that one portfolio corresponds to one clustering and is composed of 5 big assets (which are the 5 clusters) [*in this notebook, we first fix the number of clusters to be equal to 5*]

To have an idea of how we built the clustering_return() function, see the shape of cluter_data (which corresponds to the return data associated with the tickers in the cluster) and the weights of the stocks in this cluster 

In [17]:
cluster_data = data.loc[cluster]
print(cluster_data.shape)
print(weights_gaussian.shape)

(236, 230)
(1, 236)


#### **Remarks**:

- We see that the to get the return of each clusters, we have to multiply each columns of the cluster_data dataframe by the weights of the stocks (take the transpose of the line dataframe).

- The **clustering_return** function applies to a **clustering** i.e. one column of the Y_symbol dataframe (result of the cluster_composition() funciton). 

In [18]:
## We take as an example the first clustering of the Y_symbol dataframe
clustering_composition_1 = pd.DataFrame(Y_symbol.iloc[:, 0])
clustering_centroids_1 = pd.DataFrame(C.iloc[:, 0])

cluster_1 = clustering_composition_1.iloc[0, 0]
centroid_1 = clustering_centroids_1.iloc[0, 0]

In [19]:
module1.clustering_return(clustering_composition_1, clustering_centroids_1, data)

Clustering n°1
Cluster 1  [0.002464213208415318, 0.014715028989846569, -...
Cluster 2  [0.01754677984630551, -0.0003765881212793948, ...
Cluster 3  [0.015486051877961548, 0.0022706760611838177, ...
Cluster 4  [-0.007000515490829761, -0.01755430852926245, ...
Cluster 5  [0.027581074759089483, 0.022065872782772257, -...

In [38]:
Z = module1.clustering_return(Y_symbol, C, data)
Z

Clustering n°1  \
Cluster 1  [0.002464213208415318, 0.014715028989846569, -...   
Cluster 2  [0.01754677984630551, -0.0003765881212793948, ...   
Cluster 3  [0.015486051877961548, 0.0022706760611838177, ...   
Cluster 4  [-0.007000515490829761, -0.01755430852926245, ...   
Cluster 5  [0.027581074759089483, 0.022065872782772257, -...   

                                              Clustering n°2  \
Cluster 1  [-0.015961079923144984, -0.010634553189856582,...   
Cluster 2  [0.013652436197485263, 0.001988708440525629, -...   
Cluster 3  [0.0006021163573665361, 0.003295409557186626, ...   
Cluster 4  [-0.0008337244724244238, -0.015504623155711912...   
Cluster 5  [0.02667074604648615, 0.017499020013748538, -0...   

                                              Clustering n°3  \
Cluster 1  [0.00020893057564944033, 0.00831196292134567, ...   
Cluster 2  [0.016375346446535072, 0.004650578233291859, -...   
Cluster 3  [0.0005648307575109652, 0.0033482965911414454,...   
Cluster 4  [-0.005283302637740595, -0.01686469534850027, ...   
Cluster 5  [0.02742904332685718, 0.021562428551687805, -0...   

                                              Clustering n°4  \
Cluster 1  [0.0012577844738105152, 0.009535728089612292, ...   
Cluster 2  [0.007743566690972065, -0.047453427794875, -0....   
Cluster 3  [-0.002868954597134851, 0.0030538135135224076,...   
Cluster 4  [-0.00021875465892884593, 0.008960945540722876...   
Cluster 5  [0.024653915573803243, 0.01689282773461742, -0...   

                                              Clustering n°5  \
Cluster 1  [-0.0020235546434611167, 0.011835002814390493,...   
Cluster 2  [0.019499970057381594, -0.0022308706984730054,...   
Cluster 3  [-0.013413410346581634, -0.028601581586053935,...   
Cluster 4  [-0.015765420456627575, -0.007219270057129161,...   
Cluster 5  [0.037842947899023135, 0.0335931972565851, -0....   

                                              Clustering n°6  \
Cluster 1  [-0.012647233779140675, 0.005029812553833496, ...   
Cluster 2  [-0.009727851991110445, -0.021293446468710012,...   
Cluster 3  [-0.0019421288195388, 0.003381653899615703, 0....   
Cluster 4  [0.014032731772693506, 0.010824534004232587, -...   
Cluster 5  [0.028404816722818315, 0.02260931263177546, -0...   

                                              Clustering n°7  \
Cluster 1  [-0.005115475665644524, 0.008737873564470817, ...   
Cluster 2  [0.0497624837703518, -0.006974836664617619, -0...   
Cluster 3  [-0.011759644719295099, -0.005747010557304732,...   
Cluster 4  [-7.636578502436815e-05, 0.00967567030383951, ...   
Cluster 5  [0.026824156474531548, 0.02025640784871168, -0...   

                                              Clustering n°8  \
Cluster 1  [-0.00841640999870093, 0.004719841708364349, -...   
Cluster 2  [0.009283382430546957, 0.008488460762106444, -...   
Cluster 3  [0.0014914974443717994, 0.006118109801197318, ...   
Cluster 4  [0.0007280352619281803, 0.010183981109991112, ...   
Cluster 5  [0.026722713737558027, 0.02075882340192696, -0...   

                                              Clustering n°9  \
Cluster 1  [-0.005577088827769814, 0.007681404347682387, ...   
Cluster 2  [0.015297251094955482, 0.006430439735358495, -...   
Cluster 3  [-0.003937354321879313, 0.002201948544897318, ...   
Cluster 4  [-0.006839375980458135, -0.027656705154544468,...   
Cluster 5  [0.02708804572194731, 0.021255694349218958, -0...   

                                             Clustering n°10  
Cluster 1  [-0.019327710796480567, -0.01627077988142058, ...  
Cluster 2  [0.013528871546123445, -0.011817512224958315, ...  
Cluster 3  [0.010279419412019332, 0.0011521500496670773, ...  
Cluster 4  [-0.0005489389557157464, 0.012593174186519967,...  
Cluster 5  [0.02722640119070873, 0.02067294888700045, -0....

In [39]:
returns = Z.iloc[:, 0]
len(returns.values)

5

In [40]:
np_returns = np.array([returns.iloc[i] for i in range(len(returns.values))])

In [41]:
np_returns.T

array([[ 0.00246421,  0.01754678,  0.01548605, -0.00700052,  0.02758107],
       [ 0.01471503, -0.00037659,  0.00227068, -0.01755431,  0.02206587],
       [-0.01891195, -0.03363215, -0.00839963, -0.03562105, -0.03470268],
       ...,
       [-0.01014273, -0.01391672,  0.00704841, -0.00180774, -0.00029462],
       [ 0.00643239,  0.01121268,  0.01403385, -0.00831259,  0.00617351],
       [ 0.016823  ,  0.02662736,  0.00420253,  0.04936655, -0.00670652]])

In [42]:
expected_returns = np.mean(np_returns.T, axis=0)
cov_matrix = np.cov(np_returns.T, rowvar=False)

In [43]:
expected_returns 

array([-0.00010188, -0.0005411 ,  0.00033631, -0.00014543,  0.00154747])

In [45]:
module1.markowitz(expected_returns, cov_matrix)


ValueError: at least one of the assets must have an expected return exceeding the risk-free rate